compiler-plugin/src/main/java/io/ballerina/lib/cdc/compiler/codeaction/AbstractCdcCodeTemplate.java

```diff
@@ -0,0 +1,124 @@
+
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.NonTerminalNode;
+import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.compiler.syntax.tree.SyntaxTree;
+import io.ballerina.lib.cdc.compiler.DiagnosticCodes;
+import io.ballerina.projects.plugins.codeaction.CodeAction;
+import io.ballerina.projects.plugins.codeaction.CodeActionArgument;
+import io.ballerina.projects.plugins.codeaction.CodeActionContext;
+import io.ballerina.projects.plugins.codeaction.CodeActionExecutionContext;
+import io.ballerina.projects.plugins.codeaction.CodeActionInfo;
+import io.ballerina.projects.plugins.codeaction.DocumentEdit;
+import io.ballerina.tools.diagnostics.Diagnostic;
+import io.ballerina.tools.text.LineRange;
+import io.ballerina.tools.text.TextDocumentChange;
+import io.ballerina.tools.text.TextEdit;
+import io.ballerina.tools.text.TextRange;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import static io.ballerina.lib.cdc.compiler.Constants.IS_POSTGRES_LISTENER;
+import static io.ballerina.lib.cdc.compiler.Constants.NODE_LOCATION;
+import static io.ballerina.lib.cdc.compiler.Utils.findNode;
+
+/**
+ * Abstract class for CDC code templates to share common functionality.
+ */
+public abstract class AbstractCdcCodeTemplate implements CodeAction {
+
+    @Override
+    public List<String> supportedDiagnosticCodes() {
+        return List.of(
+                DiagnosticCodes.EMPTY_SERVICE.getCode(),
+                DiagnosticCodes.EMPTY_SERVICE_POSTGRESQL.getCode()
+        );
+    }
+
+    @Override
+    public Optional<CodeActionInfo> codeActionInfo(CodeActionContext codeActionContext) {
+        Diagnostic diagnostic = codeActionContext.diagnostic();
+        if (diagnostic.location() == null) {
+            return Optional.empty();
+        }
+        CodeActionArgument locationArg = CodeActionArgument.from(NODE_LOCATION, diagnostic.location().lineRange());
+        CodeActionArgument isPostgres = CodeActionArgument.from(IS_POSTGRES_LISTENER,
+                diagnostic.diagnosticInfo().code().equals(DiagnosticCodes.EMPTY_SERVICE_POSTGRESQL.getCode()));
+        return Optional.of(CodeActionInfo.from(getCodeActionDescription(), List.of(locationArg, isPostgres)));
+    }
+
+    protected abstract String getCodeActionDescription();
+
+    @Override
+    public List<DocumentEdit> execute(CodeActionExecutionContext codeActionExecutionContext) {
+        LineRange lineRange = extractLineRange(codeActionExecutionContext);
+        boolean isPostgresListener = extractIsPostgresListener(codeActionExecutionContext);
+
+        if (lineRange == null) {
+            return Collections.emptyList();
+        }
+
+        SyntaxTree syntaxTree = codeActionExecutionContext.currentDocument().syntaxTree();
+        NonTerminalNode node = findNode(syntaxTree, lineRange);
+        if (!(node instanceof ServiceDeclarationNode serviceDeclarationNode)) {
```

Comment:
No need to declare a variable right ?
```suggestion
        if (!(node instanceof ServiceDeclarationNode)) {
```

Reply:
It is used later on in the logic. Ln 87